doc/handbook/deployment-and-configuration.adoc
== Deployment and configuration

=== Requirements

Eventline is written in Go and is distributed as a statically linked
executable for various platforms.

The following platforms are officially supported for production:

- Linux x86_64.
- FreeBSD x86_64.

Binaries are also released on a best effort basis for the following platforms:

- Mac OS x86_64.
- Mac OS arm64 (M1).

Eventline uses a https://www.postgresql.org[PostgreSQL database] version 14 or
higher; the https://www.postgresql.org/docs/current/pgcrypto.html[pgcrypto]
extension must be installed. It does not require local filesystem storage.

Eventline can also send metrics to an https://www.influxdata.com[InfluxDB]
server.

=== Installation

==== Binary tarballs

Eventline is released primarily as a set of binary tarballs. Each tarball
contains everything required to run the program:

* Both the `eventline` and the `evcli` executables.
* The Eventline license.
* The `data` directory containing web assets, database migrations and template
  files.
* This handbook as a HTML file with associated images and javascript code.

While you can run Eventline directly from the tarball directory, you will want
something more practical for production. As such, binary tarballs are
generally only used by people building packages or images themselves.

==== Stable and experimental builds

Stable builds are derived from Git revisions tagged as
`v.<major>.<minor>.<patch>`. Experimental builds are derived from any other
Git revisions.

Experimental packages are made available for various platforms on a regular
basis. Their goal is to allow users to try new features before they are
released in a stable build.

CAUTION: Experimental builds are inherently less reliable than stable builds
and must not be used in production.

[#freebsd-package]
==== FreeBSD package

FreeBSD packages are available in the Exograd pkg repository. In order to
install them, you first need to add the repository as a pkg source:

[source,sh]
----
curl -sSfL -o /usr/local/etc/pkg/repos/exograd-public.conf \
     https://pkg.exograd.com/public/freebsd/exograd.conf
----

Then you can update the package index and install Eventline:

[source,sh]
----
pkg update
pkg install eventline
----

The repository provides both `eventline` and `eventline-experimental`.

[#ubuntu-package]
==== Ubuntu package

Ubuntu packages are available in the Exograd APT repository. In order to
install them, you first need to add the repository as a source for APT:

[source,sh]
----
curl -sSfL -o /etc/apt/sources.list.d/exograd-public.list \
     https://pkg.exograd.com/public/ubuntu/exograd.list
----

NOTE: The package repository is only available over HTTPS for security. You
will need to install the `ca-certificates` packages.

Then you can update the package index and install Eventline:

[source,sh]
----
apt-get update
apt-get install eventline
----

The repository provides both `eventline` and `eventline-experimental`.

===== Docker

If you intend to use the `docker` runner, you will need to install Docker. We
recommend following the
https://docs.docker.com/engine/install/ubuntu/[official Docker documentation].

You will also have to add the `eventline` user to the `docker` group to give
it access to the Docker socket:
[source,sh]
----
usermod -a -G docker eventline
----

==== Archlinux package

_Coming soon_.

==== Docker image

The `exograd/eventline` image is available on
https://hub.docker.com[DockerHub]. Versioning is the same as Eventline: for
example, Eventline 0.9.0 is available with `exograd/eventline:0.9.0`.

While the `latest` tag usually points to the latest available release, it is
strongly recommended not to use it and to always use a specific version tag to
avoid unexpected updates.

[#license-management]
===== License management

Eventline Pro requires a license allowing a client to use the product.

Each client is attributed credentials provided as two values, name and secret.
These values must be stored in the configuration file. See the
<<configuration-specification-pro,configuration specification>> for more
information.

Eventline Pro automatically uses these credentials to authenticate the Exograd
license server and download a license file.

The license is renewed automatically at startup, and on a regular basis
afterward.

[#configuration]
===== Configuration

When running in this Docker image, Eventline can be configured two ways.

One way is to mount a configuration file of your choice in the container at
`/etc/eventline/eventline.yaml`. This gives you maximum flexibility.

An over way is to rely on the embedded configuration file which uses
environment variables for configuration. The following variables are
available:

`EVENTLINE_WEB_HTTP_SERVER_URI` (optional, default to `localhost:8087`) :: The
value to use for the `web_http_server_uri` setting.

`EVENTLINE_PG_URI` (optional) :: The URI of the PostgreSQL server. The default
value is `postgres://eventline:eventline@localhost:5432/eventline`.

`EVENTLINE_ENCRYPTION_KEY` :: The value to use for the `encryption_key`
setting.

`EVENTLINE_CONNECTORS_GITHUB_WEBHOOK_SECRET` :: The value of the
`webhook_secret` setting for the `github` connector. Setting this variable
automatically enable the connector.

`EVENTLINE_MAX_PARALLEL_JOB_EXECUTIONS` :: The value to use for the
`max_parallel_job_executions` setting.

`EVENTLINE_JOB_EXECUTION_RETENTION` :: The value to use for the
`job_execution_retention` setting.

`EVENTLINE_SESSION_RETENTION` :: The value to use for the `session_retention`
setting.

`EVENTLINE_NOTIFICATIONS_SMTP_SERVER_ADDRESS` (optional, default to `localhost:25`) ::
The address of the SMTP server to use for notifications.

`EVENTLINE_NOTIFICATIONS_SMPT_SERVER_USERNAME` (optional) :: The username to
use to authenticate to the SMTP server.

`EVENTLINE_NOTIFICATIONS_SMPT_SERVER_PASSWORD` (optional) :: The password to
use to authenticate to the SMTP server.

`EVENTLINE_NOTIFICATIONS_FROM_ADDRESS` (optional) :: The email address to use in the
`From` header field.

`EVENTLINE_NOTIFICATIONS_SUBJECT_PREFIX` (optional) :: A character string to
use as prefix for the `Subject` header field.

`EVENTLINE_NOTIFICATIONS_SIGNATURE` :: A character string to insert as
signature at the end of all emails.

See the <<configuration-specification,configuration specification>> for more
information about settings.

===== Building your own image

The `exograd/eventline` is provided for convenience. For production use, it is
advised to build your own image. This allows you to follow the conventions
used by your organization and gives you total control on the base system,
environment and configuration.

Feel free to start from the default
https://github.com/exograd/eventline/blob/master/Dockerfile[Dockerfile] or to
write your own from scratch.

==== Helm chart

_Coming soon._

=== Configuration

==== Configuration file

Eventline uses a configuration file whose path is provided with the `-c`
command line option.

CAUTION: The configuration file contains the global encryption key used to
secure storage of sensitive information in the database. You must make sure
that the UNIX user executing Eventline is the only user able to read the
configuration file. Alternatively, you can use templating and environment
variables to provide sensitive settings.

==== Templating

The configuration file is treated as a template using the
https://pkg.go.dev/text/template[Go template format]. Templating currently
supports the following functions:

`env <name>` :: Return the value of the `<name>` environment variable.

.Example:
[source,yaml]
----
data_directory: "/usr/share/eventline"

encryption_key: {{ env "EVENTLINE_ENCRYPTION_KEY" }}

pg:
  uri: {{ env "EVENTLINE_PG_URI" }}
----

At startup, Eventline loads the configuration file, renders it and then parses
it as a YAML document.

[#configuration-specification]
==== Specification

A configuration file is an object containing the following fields:

`logger` (optional object) :: The configuration of the logger used to print
information and errors. The default value is:
+
[source,yaml]
----
backend_type: "terminal"
terminal_backend:
  color: true
  domain_width: 32
----

`data_directory` (optional string, default to `data`) :: The path of the
directory containing Eventline data files.

`api_http_server` (optional object) :: the HTTP server configuration of the
API interface. The default value is:
+
[source,yaml]
----
address: "localhost:8085"
----

`web_http_server` (optional object) :: the HTTP server configuration of the
web interface. The default value is:
+
[source,yaml]
----
address: "localhost:8087"
----

`pg` (optional object) :: The configuration of the PostgreSQL server.

`encryption_key` (string) :: The global encryption key used to encrypt
sensitive information in the database. The key must be a 32 byte AES key
encoded using https://en.wikipedia.org/wiki/Base64[Base64]. You can generate a
key using OpenSSL:
+
----
openssl rand -base64 32
----

`web_http_server_uri` (optional string, default to `http://localhost:8087`) ::
The URI which can be used to access the Eventline web interface from outside
of the server. This URI will be used to generate webhook URIs among other
thing.

`insecure_http_cookies` (optional boolean, default to `false`) :: If true, do
not set the secure attribute for HTTP cookies sent by the web HTTP server.
This allows Eventline to be used over HTTP.
+
CAUTION: Using Eventline without HTTPS is fundamentally insecure: anyone able
to inspect network traffic can obtain critical information transferred between
the web browser and Eventline. Do not do it.

`connectors` (optional object) :: The configuration of each connector. Refer
to the connector documentation for the settings available for each connector.

`max_parallel_job_executions` (optional integer) :: If set, the maximum number
of jobs which can run in parallel for the entire platform.

`job_execution_retention` (optional integer) :: If set, a number of days after
which old job executions will be deleted. Note that changing this setting will
not affect job executions which have already been terminated.

`job_execution_refresh_interval` (optional integer, default: 10) :: The number
of seconds between two job execution refresh. See
<<job-execution-timeout,execution documentation>> for more information on the
refresh process.

`job_execution_timeout` (optional integer, default: 120) :: The number of
seconds without refresh after which a job is considered abandonned. See
<<job-execution-timeout,execution documentation>> for more information on the
refresh process.

`session_retention` (optional integer) :: If set, a number of days after which
sessions will be deleted.

`allowed_runners` (optional string array) :: If set, a list of the runners
which can be used in submitted jobs. Jobs using other runners will be rejected
during deployment.

`runners` (optional object) :: The configuration of each runner. Refer to the
<<chapter-runners,runner documentation>> for the settings available for each
runner.

`notifications` (optional object) :: The configuration of the email
notification system. The default value is:
+
[source,yaml]
----
smtp_server:
  address: "localhost:25"
from_address: "no-reply@localhost"
subject_prefix: "[eventline] "
signature: "This email is a notification sent by the Eventline job scheduling software."
----

`pro` (optional object) :: Configuration specific to Eventline Pro. Ignored
for the open source version.

===== HTTP server specification

The configuration of a HTTP server is an object containing the following
fields:

`address` (optional string, default to `localhost:8080`) :: The address to
listen on as a `<host>:<port>` string.

`tls` (optional object) :: If set, use TLS for the connection. The object
contains the following fields:

    `certificate` (string) ::: The path of the TLS certificate file.

    `private_key` (string) ::: The path of the TLS private key.

===== PostgreSQL specification

The configuration of the PostgreSQL server is an object containing the
following fields:

`uri` (optional string) :: The URI of the PostgreSQL server. The default value
is `postgres://eventline:eventline@localhost:5432/eventline`.

===== Notifications specification

The configuration for the notification system is an object containing the
following fields:

`smtp_server` (optional object) :: The configuration of the SMTP server to use
when sending emails.

`from_address` (optional string) :: The email address to use in the `From`
header field.

`subject_prefix` (optional string) :: A character string to use as prefix for
the `Subject` header field.

`signature` (optional string) :: A character string to insert as signature at
the end of all emails.

`allowed_domains` (optional string array) :: A list of domains which can used
in notification email addresses. If the array is empty, all domains are
allowed.
+
TIP: the `allowed_domains` setting is useful to make sure you that all
notifications are sent to email addresses you control: this way you can limit
the chances of losing notifications due to a mistake, and you guarantee that
you will not send emails to someone out of your organization.

===== SMTP server specification

The configuration of the SMTP server is an object containing the following
fields:

`address` (optional string, default to `localhost:25` :: The address of the
server using the `<host>:<port>` format.

`username` (optional string) :: The username to use for authentication.

`password` (optional string) :: The password to use for authentication.

[#configuration-specification-pro]
===== Eventline Pro specification

The Eventline Pro configuration is an object containing the following fields:

`license_path` (string) :: The path of the license file.

`license_name` (string) :: The name of the license, used to authenticate
against the license server.

`license_secret` (string) :: The secret value used to authenticate against the
license server.

See documentation on <<license-management,license management>> for more
information.


CHANGELOG.md
# Changelog
## v1.1.0
_Work in progress._

### Breaking changes
- In the logger configuration, the `backend` member is replaced by either
  `terminal_backend` or `json_backend` depending on backend type configured.
  For example:
  ```yaml
  logger:
    backend_type: "terminal"
    backend:
      color: true
  ```
  Is now:
  ```yaml
  logger:
    backend_type: "terminal"
    terminal_backend:
      color: true
  ```

### Features
- Add the possibility to rename jobs. Since jobs are primarily identified by
  their name, deploying a job with a different name creates a new job instead
  of renaming the existing one. This new feature makes it possible to rename a
  job, for example to archive it.
- Add a `/jobs/id/:id/rename` API route.
- Add a `rename-job` evcli command.

## v1.0.8
### Bug fixes
- Fix login for users who have no current project when the "main" project has
  been deleted.
- Fix invalid docker `ContainerWait` parameter for recent Docker versions.

## v1.0.7
### Misc
- Truncate the output of all step executions on the web interface. The current
  limit is hardcoded to 1MB. Very large outputs cause performance issues both
  on the server and in the web browser.
- Add buffering to runner output capture to avoid overloading the database for
  jobs producing massive amounts of data. See the associated commit for more
  information.

## v1.0.6
### Bug fixes
- Fix validation for the subscription parameters of the `time` connector.

## v1.0.5
### Bug fixes
- Fix repository URI for evcli update checks.

## v1.0.4
### Bug fixes
- Fix subscription handling during the update of a job whose trigger is being
  removed.
- Fix event processing when the job has been modified and does not have a
  trigger anymore.

## v1.0.3
### Bug fixes
- Fix the leak of the client stream socket in the Docker runner.

## v1.0.2
### Bug fixes
- Fix terminal rendering so that text using black foreground remains visible.

## v1.0.1
Thanks to Adyxax for his help!

### Bug fixes
- Fix missing argument in error message in evcli.
- Recreate the subscription if the identity has changed during a job update.
- Ensure terminal restoration on login error in evcli.
- Fix ssh runner termination when the connection was never established.
- Fix support for ssh jobs with multiple users on the same host.

## v1.0.0
### Features
- Add pagination support for the `list-projects` evcli command.
- Add a `--wait` option to the `execute-job` evcli command which monitors
  execution, print status changes and wait for execution to finish before
  exiting.
- Add a `--fail` option to the `execute-job` evcli command to exit with status
  1 if job execution fails.
- Render [ANSI escape
  sequences](https://en.wikipedia.org/wiki/ANSI_escape_code) in execution
  output data on the web interface.
- `github/oauth2` identities can now be used as identity for the `docker`
  runner.
- Add support for deletion of old job executions based on the `job_retention`
  setting and the `retention` job field.
- Add support for deletion of old sessions based on the `session_retention`
  setting.
- Inject job parameters as files in `$EVENTLINE_DIR/parameters`.
- Add a notification setting to allow a specific list of email address
  domains.
- Add the `job-execution-watcher` worker to detect and stop jobs which have
  timed out, i.e. executions which have not been refreshed for some time. Job
  executions are now refreshed regularly (the interval is controlled by the
  `job_execution_refresh_interval` setting). The timeout duration is
  controlled by the `job_execution_timeout` setting.
- Add a `--validate-cfg` command line flag to exit after configuration
  validation but before starting the service.
- Add an `allowed_runners` setting to provide a list of the runners allowed in
  submitted jobs.
- Add a `generic/gpg_key` identity to store GPG keys.
- Add a new `ssh` runner to execute remote jobs.
- Add support for mount points to the `docker` runner.

### Bug fixes
- Fix job pagination in evcli.
- Fix the Docker image so that evcli can be executed inside.
- Always provide `EVENTLINE_DIR` as an absolute path.
- Fix incorrect validation of the `tls` field in http server configuration.
- Fix immediate session deletion issue when session retention is not
  configured or equal to zero.
- Fix incorrect validation of the influxdb client configuration.
- Fix initialization so that the program exits when HTTP server initialization
  fails.
- Fix configuration validation so that it fails when there is no encryption
  key.
- Fix file permissions when executing in a container as a non-root user.
- Add missing dependencies to the Systemd unit file.
- Fix filter serialization.
- Update the last use time of all identities injected in the job during
  execution.
- Fix decoding of validation errors with no data in evcli.
- Interrupt execution on job abortion.

### Misc
- Use the default monospace font of the web browser instead of serving a half
  megabyte file.
- Disable color for logging if the error output stream is not a character
  device.
- Improve validation of the configuration file.
- Add index to prevent potential performance issue during project deletion.
- Rename settings:
  - `max_parallel_jobs` to `max_parallel_job_executions`.
  - `job_retention` to `job_execution_retention`.
- Do not create events for periodic timer ticks which occurred when the server
  was down.
- The job id is now mandatory in event objects.
- Store the file containing the last build id check date in `$HOME/.evcli` for
  consistency.
- Set `HOME` for the local runner.

## v0.9.1
### Bug fixes
- Remove initial blank line(s) in notification emails.
- Remove invalid expiration date for `github/oauth2` identities.
- Fix the abort and restart buttons on job execution pages.

### Misc
- Use a stable tag in the docker-compose setup.

## v0.9.0
First public release.


pkg/service/cfg.go
package service

import (
	"encoding/json"

	"github.com/exograd/eventline/pkg/cryptoutils"
	"github.com/exograd/eventline/pkg/eventline"
	"github.com/galdor/go-ejson"
	"github.com/galdor/go-log"
	"github.com/galdor/go-service/pkg/influx"
	"github.com/galdor/go-service/pkg/pg"
	goservice "github.com/galdor/go-service/pkg/service"
	"github.com/galdor/go-service/pkg/shttp"
)

type ServiceCfg struct {
	Logger *log.LoggerCfg `json:"logger"`

	ServiceAPI *goservice.ServiceAPICfg `json:"service_api"`

	DataDirectory string `json:"data_directory"`

	APIHTTPServer *shttp.ServerCfg `json:"api_http_server"`
	WebHTTPServer *shttp.ServerCfg `json:"web_http_server"`

	Influx *influx.ClientCfg `json:"influx"`

	Pg *pg.ClientCfg `json:"pg"`

	EncryptionKey cryptoutils.AES256Key `json:"encryption_key"`

	WebHTTPServerURI    string `json:"web_http_server_uri"`
	InsecureHTTPCookies bool   `json:"insecure_http_cookies"`

	Connectors map[string]json.RawMessage `json:"connectors"`

	Workers map[string]eventline.WorkerCfg `json:"workers"`

	MaxParallelJobExecutions    int `json:"max_parallel_job_executions"`
	JobExecutionRetention       int `json:"job_execution_retention"`        // days
	JobExecutionRefreshInterval int `json:"job_execution_refresh_interval"` // seconds
	JobExecutionTimeout         int `json:"job_execution_timeout"`          // seconds

	SessionRetention int `json:"session_retention"` // days

	AllowedRunners []string                   `json:"allowed_runners"`
	Runners        map[string]json.RawMessage `json:"runners"`

	Notifications *NotificationsCfg `json:"notifications"`

	ProCfg ejson.Validatable `json:"pro"`
}

func DefaultServiceCfg() ServiceCfg {
	logger := &log.LoggerCfg{
		BackendType: "terminal",
		TerminalBackend: &log.TerminalBackendCfg{
			Color:       true,
			DomainWidth: 32,
		},
	}

	return ServiceCfg{
		Logger: logger,

		DataDirectory: "data",

		APIHTTPServer: &shttp.ServerCfg{
			Address: "localhost:8085",
		},
		WebHTTPServer: &shttp.ServerCfg{
			Address: "localhost:8087",
		},

		Pg: &pg.ClientCfg{
			URI:         "postgres://eventline:eventline@localhost:5432/eventline",
			SchemaNames: []string{"eventline"},
		},

		WebHTTPServerURI: "http://localhost:8087",

		JobExecutionRefreshInterval: 10,
		JobExecutionTimeout:         120,

		Notifications: DefaultNotificationsCfg(),
	}
}

func (cfg *ServiceCfg) Check(v *ejson.Validator, s *Service) {
	// Note that some fields are optional in the documentation but mandatory
	// here. These values are set in the default configuration: they must be
	// provided for Eventline to work, but we define reasonable default values
	// so that the user does not have to set them in most cases.
	//
	// Also, we do not check the value of the encryption key, since it is
	// validated by its UnmarshalJSON method. We still have to check that it
	// is present, since UnmarshalJSON will not be called if the field is not
	// set.
	//
	// Finally, connectors, workers and runners are currently handled later in
	// the initialization phase. This is clearly something we could improve in
	// the future.

	v.CheckObject("logger", cfg.Logger)
	v.CheckOptionalObject("service_api", cfg.ServiceAPI)

	v.CheckStringNotEmpty("data_directory", cfg.DataDirectory)

	v.CheckObject("api_http_server", cfg.APIHTTPServer)
	v.CheckObject("web_http_server", cfg.WebHTTPServer)

	v.CheckOptionalObject("influx", cfg.Influx)

	v.CheckObject("pg", cfg.Pg)

	v.Check("encryption_key", !cfg.EncryptionKey.IsZero(),
		"invalid_value", "missing encryption key")

	v.CheckStringURI("web_http_server_uri", cfg.WebHTTPServerURI)

	if cfg.MaxParallelJobExecutions != 0 {
		v.CheckIntMin("max_parallel_job_executions",
			cfg.MaxParallelJobExecutions, 1)
	}

	if cfg.JobExecutionRetention != 0 {
		v.CheckIntMin("job_execution_retention", cfg.JobExecutionRetention, 1)
	}

	if cfg.JobExecutionRefreshInterval != 0 {
		v.CheckIntMin("job_execution_refresh_interval",
			cfg.JobExecutionRefreshInterval, 1)
	}

	if cfg.JobExecutionTimeout != 0 {
		v.CheckIntMin("job_execution_timeout", cfg.JobExecutionTimeout, 1)
	}

	if cfg.SessionRetention != 0 {
		v.CheckIntMin("session_retention", cfg.SessionRetention, 1)
	}

	v.WithChild("allowed_runners", func() {
		for i, r := range cfg.AllowedRunners {
			v.CheckStringValue(i, r, s.runnerNames)
		}
	})

	v.CheckObject("notifications", cfg.Notifications)
}
